src/apis/web3api.py

```diff
@@ -0,0 +1,215 @@
+from src.helper_functions import get_logger
+from src.constants import (
+    ADDRESS,
+)
+
+
+class Web3API:
+    """
+    Class for fetching data from a Web3 API.
+    """
+
+    def __init__(self):
+        load_dotenv()
+        infura_key = getenv("INFURA_KEY")
+        self.url = f"https://mainnet.infura.io/v3/{infura_key}"
+        self.web_3 = Web3(Web3.HTTPProvider(self.url))
+        self.contract = self.web_3.eth.contract(
+            address=Address(HexBytes(ADDRESS)), abi=gpv2Abi
+        )
+        self.logger = get_logger()
+
+    def get_current_block_number(self) -> Optional[int]:
+        """
+        Function that returns the current block number
+        """
+        try:
+            return int(self.web_3.eth.block_number)
+        except ValueError as err:
+            self.logger.error("Error while fetching block number: %s", err)
+            return None
+
+    def get_tx_hashes_by_block(self, start_block: int, end_block: int) -> List[str]:
+        """
+        Function filters hashes by contract address, and block ranges
+        """
+        filter_criteria = {
+            "fromBlock": int(start_block),
+            "toBlock": int(end_block),
+            "address": ADDRESS,
+            "topics": [
+                "0xa07a543ab8a018198e99ca0184c93fe9050a79400a0a723441f84de1d972cc17"
+            ],
+        }
+
+        try:
+            transactions = self.web_3.eth.filter(filter_criteria).get_all_entries()  # type: ignore
+        except ValueError as except_err:
+            self.logger.error(
+                "ValueError while fetching hashes: %s.",
+                str(except_err),
+            )
+            transactions = []
+
+        settlement_hashes_list = []
+        # transactions may have repeating hashes, since even event logs are filtered
+        # therefore, check if hash has already been added to the list
+        for transaction in transactions:
+            tx_hash = (transaction["transactionHash"]).hex()
+            if tx_hash not in settlement_hashes_list:
+                settlement_hashes_list.append(tx_hash)
+        return settlement_hashes_list
+
+    def get_transaction(self, tx_hash: str) -> Optional[TxData]:
+        """
+        Takes settlement hash as input, returns transaction data.
+        """
+        try:
+            transaction = self.web_3.eth.get_transaction(HexStr(tx_hash))
+        except ValueError as err:
+            self.logger.debug(
+                "Error while fetching transaction: %s.",
+                str(err),
+            )
+            transaction = None
+
+        return transaction
+
+    def get_receipt(self, tx_hash: str) -> Optional[TxReceipt]:
+        """
+        Get the receipt of a transaction from the transaction hash.
+        This is used to obtain the gas used for the transaction.
+        """
+        try:
+            receipt = self.web_3.eth.wait_for_transaction_receipt(HexStr(tx_hash))
```

Comment:
Since we are using Web3 Library and conforming to their types, you could even go as far as putting these types directly into the function signature like:

```py
def get_receipt(self, tx_hash: HexStr) -> Optional[TxReceipt]:
```

Then the end user would be expected to pass ready information to the function.

Reply:
This might also be a good place to drill down into the web3 library (to see how `TxReceipt` is constructed from what is likely a json response). Might give some hints as to how one might eliminate all of these `dict[str, Any]` cropping up.